packages/scripts/scripts/check-licenses.js

```diff
@@ -170,36 +169,46 @@ const checkLicense = ( allowedLicense, licenseType ) => {
 	);
 };
 
-/**
- * Returns true if the given module path is not to be ignored for consideration
- * in license validation, or false otherwise.
- *
- * @param {string} moduleName Module path.
- *
- * @return {boolean} Whether module path is not to be ignored.
- */
-const isNotIgnoredModule = ( moduleName ) =>
-	! ignored.some( ( ignoredItem ) =>
-		// `moduleName` is a file path to the module directory. Assume CLI arg
-		// is passed as basename of package (directory(s) after node_modules).
-		// Prefix with sep to avoid false-positives on prefixing variations.
-		moduleName.endsWith( sep + ignoredItem )
-	);
-
 // Use `npm ls` to grab a list of all the packages.
-const child = spawn.sync( 'npm', [
-	'ls',
-	'--parseable',
-	...( prod ? [ '--prod' ] : [] ),
-	...( dev ? [ '--dev' ] : [] ),
-] );
-
-const modules = child.stdout
-	.toString()
-	.split( '\n' )
-	.filter( isNotIgnoredModule );
-
-modules.forEach( ( path ) => {
+const child = spawn.sync(
+	'npm',
+	[
+		'ls',
+		'--json',
+		'--long',
+		...( prod ? [ '--prod' ] : [] ),
+		...( dev ? [ '--dev' ] : [] ),
+	],
+	{ maxBuffer: 1024 * 1024 * 100 } // output size for prod is ~21 MB and dev is ~76 MB
+);
+
+const result = JSON.parse( child.stdout.toString() );
+
+const topLevelDeps = result.dependencies;
+
+function traverseDepTree( deps ) {
+	for ( const key in deps ) {
+		const dep = deps[ key ];
+
+		if ( ignored.includes( dep.name ) ) {
+			return;
+		}
+
+		if ( ! dep.hasOwnProperty( 'path' ) ) {
+			// throw Error( '??' );
+		} else {
+			checkDepLicense( dep.path );
```

Comment:
I guess there's going to be a lot of duplicate checks since dependencies are shared. Could we cache that maybe?
```
const scannedPackages = {};
function traverseDepTree( deps ) {
...
if ( ! scannedPackages[ dep.path ] ) {
    checkDepLicense( dep.path );
    scannedPackages[ dep.path ] = true;
}
```
Not sure if the script is slow at the moment?

Reply:
I'm seeing we use `require( filename );` a few lines below to load `package.json` files which means their content will be cached automatically by node.js, so it's not that much of an issue.

Reply:
It's taking around ~12 seconds currently on my machine. It was ~9s before these changes.
With caching it's around ~11s. I think it's not necessary to complicate the code any further with caching.